system-addon/lib/TopStoriesFeed.jsm

```diff
@@ -277,6 +297,83 @@ this.TopStoriesFeed = class TopStoriesFeed {
     this.newTabsSinceSpoc++;
   }
 
+  // Frequency caps are based on campaigns, which may include multiple spocs.
+  // We currently support two types of frequency caps:
+  // - lifetime: Indicates how many times spocs from a campaign can be shown in total
+  // - period: Indicates how many times spocs from a campaign can be shown within a period
+  //
+  // So, for example, the feed configuration below defines that for campaign 1 no more
+  // than 5 spocs can be show in total, and no more than 2 per hour.
+  // "campaign_id": 1,
+  // "caps": {
+  //  "lifetime": 5,
+  //  "campaign": {
+  //    "count": 2,
+  //    "period": 3600
+  //  }
+  // }
+  isBelowFrequencyCap(impressions, spoc) {
+    const campaignImpressions = impressions[spoc.spoc_meta.campaign_id];
+    if (!campaignImpressions) {
+      return true;
+    }
+
+    const lifeTimeCap = spoc.spoc_meta.caps && spoc.spoc_meta.caps.lifetime;
+    const campaignCap = (spoc.spoc_meta.caps && spoc.spoc_meta.caps.campaign) || {};
+
+    const lifeTimeCapExceeded = () => campaignImpressions.length >= lifeTimeCap;
+    const campaignCapExceeded = () =>
+      campaignImpressions.filter(i => (Date.now() - i) < (campaignCap.period * 1000)).length >= campaignCap.count;
+
+    return !lifeTimeCapExceeded() && !campaignCapExceeded();
+  }
+
+  updateActiveCampaigns(spocs) {
+    const activeCampaigns = new Map();
```

Comment:
Reading through this patch again, `activeCampaigns` sounds a bit confusing to me, particularly, there is a `activeCampaignIds` below. It's actually a map from tile_id to campaign_id, perhaps rename it? e.g. `mapTileToCampaign`. Heck, naming stuff is really hard ;)

Reply:
OK, I will think about the names again and update later tonight.